### DISCUSSION TITLE
Missing Attribute Reference in Agent Code

### DISCUSSION THREAD

**Discussion by bytetec** [FLAMEGPU/FLAMEGPU2#1163] [General]
Hi all
Is there a possibility to configure the references of pyflamegpu functions/attributes in the agent code, so that also code completion can be used? The missing references only happen in the agent code, in the rest of the code in works. Could it be that I installed the wheel wrongly?
I’m using PyCharm as an IDE.

![image](https://github.com/FLAMEGPU/FLAMEGPU2/assets/138150718/3898eca0-5826-466c-8020-d887f2fd93b5)

**Comment by Robadob** [comment]
> Is there a possibility to configure the references of pyflamegpu functions/attributes in the agent code, so that also code completion can be used?

Agent python does not actually exist as a typical Python package, it is transpiled to equivalent C++ syntax when you pass it to `codegen.translate()` and never executed as Python. This explains why `PyCharm` can't natively detect it's methods.

We could potentially add support for this via `.pyi` files, however the tool that we use for automatically generating the Python interface [does not currently support this](https://github.com/swig/swig/issues/735). Hence, it would likely be a rather involved manual effort to both initially build and subsequently maintain.

We've created an issue (https://github.com/FLAMEGPU/FLAMEGPU2/issues/1165) to track this feature-request, however it's unlikely to be addressed in the immediate future.

**Reply by bytetec** [reply]
Thanks for the realy quick response 👍
